### PR TITLE
🏆 Custom reward function for GRPO

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -51,22 +48,89 @@
...
+        model (`Union[str, PreTrainedModel]`):
+            Model to be trained. Can be either:
+
+            - A string, being the *model id* of a pretrained model hosted inside a model repo on huggingface.co, or
+              a path to a *directory* containing model weights saved using
+              [`~transformers.PreTrainedModel.save_pretrained`], e.g., `'./my_model_directory/'`. The model is
+              loaded using [`~transformers.AutoModelForCausalLM.from_pretrained`] with the keywork arguments
+              in `args.model_init_kwargs`.
+            - A [`~transformers.PreTrainedModel`] object. Only causal language models are supported.
+        reward_model (`Union[str, PreTrainedModel, Callable[[list, list], list[float]]`):
```
Comment (andyl98) [inline trl/trainer/grpo_trainer.py:82]: nit: I feel like the `reward_model` naming might be a bit confusing for users, i.e. if they didn't read the docstring, they might assume this has to be a LLM.

Can we rename it to sth like `reward_provider` or `reward_accessor`?

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:82]: > if they didn't read the docstring

Who the hell doesn't read the doc???

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:82]: Just kidding, what about `reward_func`? A torch module is a callable, so somehow it's a function. Maybe less confusing

**Comment by plaguss** [comment]
LGTM, but to compare against a reference answer for example we would need access to another column in the dataset, a function similar to this one:

```python
def reward_func(prompts, completions, references=None):
    """Reward function that gives higher scores to completions that are the same as the reference answer."""
    return [float(completion == reference) for completion, reference in zip(completions, references)]
```

From the R1 paper:
image

**Review by plaguss — APPROVED** [review]

**Comment by qgallouedec** [comment]
Migration guide:

```diff
  trainer = GRPOTrainer(
- reward_model=...,
+ reward_funcs=...,
- reward_processing_class=...,
+ reward_processing_classes=...,
  )
```

**Review by lewtun — APPROVED** [review]
Great feature! LGTM modulo the question of training on mixed verifiable / non-verifiable data (but that can be left for later IMO)
